### PR TITLE
Add justification checking and block height checking

### PR DESCRIPTION
**Pull request by birchmd**
## Overview
Do not accept invalid blocks into the Casper internal state

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/RHOL-376

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 1]
```diff
@@ -37,4 +37,42 @@ object Validate {
...
+        val trueParents = ProtoUtil.chooseNonConflicting(estimate, genesis, dag).map(_.blockHash)
+        val bParents    = b.header.map(_.parentsHashList).getOrElse(Seq.empty[BlockMessage])
+
+        if (bParents == trueParents)
+          true.pure[F]
+        else
+          Log[F].warn(
+            ignore(b, "block parents did not match estimate based on justification")
+          ) *> false.pure[F]
+      })
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: Refactoring to remove the block.pure[F].flatMap construct:

```
  def parents[F[_]: Monad: Log](block: BlockMessage,
                                genesis: BlockMessage,
                                dag: BlockDag): F[Boolean] = {
    def listedParents(block: BlockMessage) =
      block.header.map(_.parentsHashList).getOrElse(Seq.empty[BlockMessage])
    def computedParents(block: BlockMessage, genesis: BlockMessage, dag: BlockDag) = {
      val latestMessages = block.justifications.map {
        case Justification(v, hash) => v -> hash
      }.toMap
      val viewDag  = dag.copy(latestMessages = latestMessages)
      val estimate = Estimator.tips(viewDag, genesis)
      ProtoUtil.chooseNonConflicting(estimate, genesis, dag).map(_.blockHash)
    }

    for {
      result <- if (listedParents(block) == computedParents(block, genesis, dag))
                 true.pure[F]
               else
                 Log[F].warn(
                   ignore(block, "block parents did not match estimate based on justification")
                 ) *> false.pure[F]
    } yield result
  }
```

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 2]
```diff
@@ -37,4 +37,42 @@ object Validate {
...
+          ProtoUtil.parents(b).map(dag.blockLookup andThen ProtoUtil.blockNumber).max
+        val number = ProtoUtil.blockNumber(b)
+        if (parentNumber + 1 != number) {
+          Log[F].warn(
+            ignore(b, s"block number $number is not one more than parent number $parentNumber.")
+          ) *> false.pure[F]
+        } else {
+          true.pure[F]
+        }
+      })
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: I think the parentNumber should just be the parent's mainChain + 1.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: In either case here is a refactoring without block.pure[F].flatMap

```
  def blockNumber[F[_]: Monad: Log](block: BlockMessage, dag: BlockDag): F[Boolean] = {
    def parentNumber(block: BlockMessage, dag: BlockDag) =
      ProtoUtil.parents(block).map(dag.blockLookup andThen ProtoUtil.blockNumber).max
    def number(block: BlockMessage) = ProtoUtil.blockNumber(block)

    for {
      result <- if (parentNumber(block, dag) + 1 != number(block)) {
                 Log[F].warn(
                   ignore(
                     block,
                     s"block number ${number(block)} is not one more than parent number ${parentNumber(block, dag)}.")
                 ) *> false.pure[F]
               } else {
                 true.pure[F]
               }
    } yield result
  }
```

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 3]
```diff
@@ -190,74 +183,78 @@ sealed abstract class MultiParentCasperInstances {
...
-            //TODO: Ask peers for missing parents/justifications of blocks
-            case f @ Some(false) => Monad[F].pure[Option[Boolean]](f)
-
-            case _ =>
-              Log[F].info(
-                s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(block.blockHash)}") *>
-                Monad[F].pure[Option[Boolean]](None)
-          }
+        }
+      }.flatMap {
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/Casper.scala]: Can we stick to the for yield syntax? I just find it more elegant and I think its more consistent with @rabbitonweb 's style

```
      //invalid blocks return None
      private def attemptAdd(block: BlockMessage): F[Option[Boolean]] =
        for {
          success <- {
            val hash     = block.blockHash
            val bParents = parents(block)

            if (bParents.exists(p => !_blockDag.get.blockLookup.contains(p))) {
              //Cannot add a block if not all parents are known,
              //but can re-try in the future
              Monad[F].pure[Option[Boolean]](Some(false))
            } else if (block.justifications.exists(j =>
                         !_blockDag.get.blockLookup.contains(j.latestBlockHash))) {
              //Cannot add a block if not all justifications are known,
              //but can re-try in the future
              Monad[F].pure[Option[Boolean]](Some(false))
            } else {
              val Right((maybeCheckPoint, _)) =
                checkpoints.mapAndUpdate[(Option[Checkpoint], Map[ByteString, Checkpoint])](
                  InterpreterUtil.validateBlockCheckpoint(
                    block,
                    genesis,
                    _blockDag.get,
                    storageLocation,
                    storageSize,
                    _
                  ),
                  _._2
                )

              val isValid = maybeCheckPoint.traverse(_ => {
                val dag = _blockDag.get
                for {
                  bn <- Validate.blockNumber[F](block, dag)
                  p  <- Validate.parents[F](block, genesis, dag)
                } yield bn && p
              })

              isValid.map(_.flatMap {
                case true =>
                  _blockDag.update(bd => {
                    //add current block as new child to each of its parents
                    val newChildMap = bParents.foldLeft(bd.childMap) {
                      case (acc, p) =>
                        val currChildren = acc.getOrElse(p, HashSet.empty[BlockHash])
                        acc.updated(p, currChildren + hash)
                    }

                    bd.copy(
                      blockLookup = bd.blockLookup.updated(hash, block),
                      //Assume that a non-equivocating validator must include
                      //its own latest message in the justification. Therefore,
                      //for a given validator the blocks are guaranteed to arrive in causal order.
                      latestMessages = bd.latestMessages.updated(block.sender, hash),
                      childMap = newChildMap
                    )
                  })
                  Some(true)

                case false => None
              })
            }
          }
          result <- success match {
                     case t @ Some(true) =>
                       //Add successful! Send block to peers and create tuplespace checkpoint
                       CommUtil.sendBlock[F](block) *> Monad[F].pure[Option[Boolean]](t)

                     //TODO: Ask peers for missing parents/justifications of blocks
                     case f @ Some(false) => Monad[F].pure[Option[Boolean]](f)

                     case _ =>
                       Log[F].info(
                         s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(block.blockHash)}") *>
                         Monad[F].pure[Option[Boolean]](None)
                   }
        } yield result
```

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: We probably need to break up this function soon too.

**Comment by birchmd** [comment]
@KentShikama PTAL I did a bunch of refactoring and added some unit tests.

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 4]
```diff
@@ -1,40 +1,96 @@
...
+      if (isValid) {
+        true.pure[F]
+      } else {
+        Log[F].warn(ignore(b, "signature is invalid.")) *> false.pure[F]
+      }
+    } else {
+      Log[F].warn(
+        ignore(b, s"signature algorithm ${b.sigAlgorithm} is unsupported.")
+      ) *> false.pure[F]
+    }
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: Hmm so if this is safe, what was the issue that made us cautious about wrapping everything in Capture?

File: casper/src/test/scala/coop/rchain/casper/ValidateTest.scala [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    def createChainWithValidators[F[_]: Monad: BlockDagState]: F[BlockMessage] =
+      for {
+        b0 <- createBlock[F](Seq.empty, bonds = bonds)
+        b1 <- createValidatorBlock[F](Seq(b0), Seq.empty, 0)
+        b2 <- createValidatorBlock[F](Seq(b0), Seq.empty, 1)
+        b3 <- createValidatorBlock[F](Seq(b0), Seq.empty, 2)
+        b4 <- createValidatorBlock[F](Seq(b1), Seq(b1), 0)
+        b5 <- createValidatorBlock[F](Seq(b3, b2, b1), Seq(b1, b2, b3), 1)
+        b6 <- createValidatorBlock[F](Seq(b5), Seq(b1, b4, b5), 0)
+        b7 <- createValidatorBlock[F](Seq(b4), Seq(b1, b4, b5), 1) //not heightest weight parent
```
Comment (KentShikama) [inline casper/src/test/scala/coop/rchain/casper/ValidateTest.scala]: heaviest

Comment (KentShikama) [reply casper/src/test/scala/coop/rchain/casper/ValidateTest.scala]: also score not weight

File: casper/src/test/scala/coop/rchain/casper/ValidateTest.scala [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+  it should "return true for sequential numbering" in {
+    val n     = 6
+    val chain = createChain[StateWithChain](n).runS(initState).value
+
+    (0 until n).forall(i => Validate.blockNumber[Id](chain.idToBlocks(i), chain)) should be(true)
+    log.warns should be(Nil)
+  }
+
+  "Parent validation" should "return true for proper justifications and false otherwise" in {
```
Comment (KentShikama) [inline casper/src/test/scala/coop/rchain/casper/ValidateTest.scala]: I've created a bounty task for creating a visualization for this https://rchain.atlassian.net/browse/RHOL-388 . The comments on b7 ~ b9 make it pretty obvious already.

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 4 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: Pawel said he'll teach us tomorrow unless you already have a solid understanding.

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: The point of `F` is to capture effects. So you do not need to wrap something in a `F` when it is a pure function. Here, I do some pure checks against the input values and the only effect is the logging, which is why we don't need any `F` until the end when we write the log message and return the result. What I had done wrong before was pretty niche and I over reacted by wrapping everything in `F`s. The problem there was that I was producing a "recipe" (as Pawel says) and then calling `.forever` on it, but an important check was not contained in the recipe, so the check was not repeated.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: And thanks for the explanation

**Comment by birchmd** [comment]
@KentShikama Sorry, I dismissed the review. Probably should have merged and then made a new PR... But this new commit adds some unit tests for HashSetCasper and importantly catches a bug with regard to not being able to add blocks that use unforgable names in their Rholang (because it was failing tuplespace validation).

**Pull request merged**